src/armature/Generator.ts

```diff
@@ -253,6 +253,102 @@ export class GeneratorInstance {
     }
 }
 
+type SOSMCParams = {
+    start: string;
+    sosmcDepth?: number;
+    samples?: number | ((generation: number) => number);
+    costFn: CostFn;
+    heuristicScale?: number | ((generation: number) => number);
+
+    /**
+     * For debugging, a callback can be passed in so that every sample in the final
+     * generation can be examined.
+     */
+    iterationHook?: (instances: GeneratorInstance[]) => void;
+
+    timeBudget?: number;
+};
+
+export type GeneratorStats = {
+    realTime: number;
+    cpuTime: number;
+};
+
+/**
+ * A class that controls the generation of a model using SOSMC. It schedules the optimization
+ * to be broken up into chunks that run within a given time budget per frame, allowing the UI
+ * to update. It can also be cancelled.
+ */
+export class GeneratorTask {
+    private result: Model | undefined;
+    private cancelled: boolean = false;
+    private onComplete: ((model: Model, stats: GeneratorStats) => void) | undefined;
+    private timeBudget: number = Infinity;
+    private startTime: number = new Date().getTime();
+    private cpuTime: number = 0;
+
+    /**
+     * @param {number} timeBudget How much time in seconds can be spent per frame before the rest
+     * of the work is deferred to the next frame. Note that it will only stop once it has gone OVER
+     * this time budget (hopefully just by a little bit) so it is not a hard limit.
+     */
+    constructor(iterator: IterableIterator<Model | undefined>, timeBudget?: number) {
```

Comment:
Could you add documentation for the iterator please too. If I'm being too nitty, just ignore me lol.

Reply:
I'm going to add the documentation onto `generateSOSMCIterator` since that's where this comes from. But basically it returns a `Model` when all the work is done, or `undefined` if it's still in progress.